#include "NCRLabViz.hpp"

using namespace std::chrono_literals;

NCRLabViz::NCRLabViz() : Node("ncr_lab_viz_node"), num_subscribers_(0) {
	RCLCPP_INFO(this->get_logger(), "Initializing NCR Lab Visualization Node");

	// Get parameters
    std::vector<double>temp;
    this->declare_parameter("border_x", temp);
    this->declare_parameter("border_z", temp);

    std::vector<double>border_x, border_z;
    this->get_parameter("border_x", border_x);
    this->get_parameter("border_z", border_z);

    // Enforce border size
    if (border_x.size() != 4 || border_z.size() != 4) {
        RCLCPP_ERROR(this->get_logger(), "(NCRLabViz) Border X and Y must be length 4.");
    }

    // Initialize polys
    low_border_poly_msg_.header.frame_id = high_border_poly_msg_.header.frame_id = "odom";

    // Set timestamps
    rclcpp::Time now = this->get_clock()->now();
    low_border_poly_msg_.header.stamp = now;
    high_border_poly_msg_.header.stamp = now;

    // Populate coordinates for border polys
    for (size_t i=0; i < border_x.size(); i++) {
        geometry_msgs::msg::Point32 p_low, p_high;

        p_low.x = p_high.x = border_x[i];
        p_low.z = p_high.z = border_z[i];
        p_low.y = 0.0;
        p_high.y = 2.7432;

        low_border_poly_msg_.polygon.points.push_back(p_low);
        high_border_poly_msg_.polygon.points.push_back(p_high);
    }

    // Set marker types
    // low_border_marker.type = high_border_marker.type = visualization_msgs::msg::Marker::LINE_STRIP;
    // post_marker.type = visualization_msgs::msg::Marker::CYLINDER;

    // Set colors
    // 

    // //Border is white
    // border_color.r = 1.0;
    // border_color.g = 1.0;
    // border_color.b = 1.0;
    // border_color.a = 1.0;
    // low_border_marker.color = high_border_marker.color = border_color;

    //Add one more point to complete the loop
    // geometry_msgs::msg::Point p_low, p_high;
    // p_low.x = p_high.x = border_x[0];
    // p_low.y = p_high.y = border_y[0];
    // p_low.z = 0.0;
    // p_high.z = 5.0;
    // low_border_marker.points.push_back(p_low);
    // high_border_marker.points.push_back(p_high);

    //Wait until someone subscribes before publishing
    // rclcpp::Rate loop_rate(1);
    // while (!posts_publisher_->get_subscription_count()) {
    //     loop_rate.sleep();
    // }
    
    // low_border_marker.header.frame_id = "autonomy_park";
    // low_border_marker.ns = "while_line";
    // low_border_marker.id = 0;
    // low_border_marker.type = visualization_msgs::msg::Marker::LINE_STRIP;
    // low_border_marker.action = 0;
    // low_border_marker.pose = geometry_msgs::msg::Pose();
    // low_border_marker.scale.x = 0.02;
    // low_border_marker.color.a = 1.0;
    // low_border_marker.color.b = 1.0;
    // low_border_marker.color.g = 1.0;
    // low_border_marker.color.r = 1.0;
    // auto point1 = geometry_msgs::msg::Point();
    // point1.x = 1.0;
    // point1.y = 1.0;
    // point1.z = 0.0;
    // auto point2 = geometry_msgs::msg::Point();
    // point2.x = 2.0;
    // point2.y = 2.0;
    // point2.z = 0.0;
    // low_border_marker.points.push_back(point1);
    // low_border_marker.points.push_back(point2);
    // marker_array_.markers.push_back(low_border_marker);
    // marker_array_.markers.push_back(high_border_marker);

    low_border_publisher_ = this->create_publisher<geometry_msgs::msg::PolygonStamped>("low_border", 1);
    high_border_publisher_ = this->create_publisher<geometry_msgs::msg::PolygonStamped>("high_border", 1);

    timer_ = this->create_wall_timer(1s, std::bind(&NCRLabViz::timer_callback, this));
}

void NCRLabViz::timer_callback() {
    this->low_border_publisher_->publish(low_border_poly_msg_);
    this->high_border_publisher_->publish(high_border_poly_msg_);
}